swap/src/bitcoin/wallet.rs

```diff
@@ -7,55 +7,93 @@ use crate::{
     execution_params::ExecutionParams,
 };
 use ::bitcoin::{util::psbt::PartiallySignedTransaction, Txid};
-use anyhow::{Context, Result};
+use anyhow::{anyhow, Result};
 use async_trait::async_trait;
 use backoff::{backoff::Constant as ConstantBackoff, tokio::retry};
-use bitcoin_harness::{bitcoind_rpc::PsbtBase64, BitcoindRpcApi};
-use reqwest::Url;
-use std::time::Duration;
-use tokio::time::interval;
+use bdk::{
+    blockchain::{noop_progress, Blockchain, ElectrumBlockchain},
+    electrum_client::{Client, ElectrumApi},
+    keys::GeneratableDefaultOptions,
+    FeeRate,
+};
+use reqwest::{Method, Url};
+use serde::{Deserialize, Serialize};
+use std::{path::Path, sync::Arc, time::Duration};
+use tokio::{sync::Mutex, time::interval};
+use tracing::debug;
 
-#[derive(Debug)]
 pub struct Wallet {
-    pub inner: bitcoin_harness::Wallet,
+    pub inner: Arc<Mutex<bdk::Wallet<ElectrumBlockchain, bdk::sled::Tree>>>,
     pub network: bitcoin::Network,
+    pub http_url: Url,
+    pub rpc_url: Url,
 }
 
 impl Wallet {
-    pub async fn new(name: &str, url: Url, network: bitcoin::Network) -> Result<Self> {
-        let wallet = bitcoin_harness::Wallet::new(name, url).await?;
+    pub async fn new(
+        electrum_rpc_url: Url,
+        electrum_http_url: Url,
+        network: bitcoin::Network,
+        datadir: &Path,
+    ) -> Result<Self> {
+        let client =
+            Client::new(electrum_rpc_url.as_str()).expect("Failed to init electrum rpc client");
+
+        let db = bdk::sled::open(datadir).expect("could not open sled db");
+        let db = db.open_tree("default_tree").expect("could not open tree");
+
+        // todo: make key generation configurable using a descriptor
+        let p_key = ::bitcoin::PrivateKey::generate_default()?;
+        let bdk_wallet = bdk::Wallet::new(
+            bdk::template::P2WPKH(p_key),
+            None,
+            network,
+            db,
+            ElectrumBlockchain::from(client),
+        )?;
 
         Ok(Self {
-            inner: wallet,
+            inner: Arc::new(Mutex::new(bdk_wallet)),
             network,
+            http_url: electrum_http_url,
+            rpc_url: electrum_rpc_url,
         })
     }
 
     pub async fn balance(&self) -> Result<Amount> {
-        let balance = self.inner.balance().await?;
-        Ok(balance)
+        self.inner.lock().await.sync(noop_progress(), None)?;
```

Comment:
Yeh this has been noted in the list of issues. There was mention in the bdk docs of moving sync into a seperate thread and calling it in a loop.

Reply:
I propose the following change, instead of doing this in the wallet's balance function, where it clearly does not belong and most likely will cause blocking trouble.I would suggest to move it to the caller as @thomaseizinger suggested.

For the tests I see this right before we assert on balances. For production we might want to think about it a bit more, but as a start it might be sufficient to sync before and after a swap.